### ISSUE TITLE
Missing hoisting of variables inside functions

### ISSUE THREAD

**Issue by InDIOS** [NectarJS/nectarjs#107]
Hi @adrien-thierry, i've tested some codes and i found other issue. When you have a variable declared with `var` keyword in global scope everything works fine, the problem comes when you have this code: 
```js
function some(param) {
  if (param) {
    var a = 1; // this var here throws an error at compile time
  }
  console.log(a);
}
some(1);
```
This happen because in C++ this is an error, but in javascript this is right.
You could do something like this
```js
function some(param) {
  var a; // now this var works just fine
  if (param) {
    a = 1; 
  }
  console.log(a);
}
some(1);
```
This can mimic the Javascript specs in C++. This time i owe you a suggestion 'cause this can comes with changes in several files. I think that `FunctionDeclaration` visitor hasn't this functionality implemented. You can start in it.
Compile time Error:
```
C:\Users\Robert\Desktop\.nectar\wk6sy\test.cpp: In lambda function:
C:\Users\Robert\Desktop\.nectar\wk6sy\test.cpp:289:18: error: 'a' was not declared in this scope
  289 |   console["log"](a);
      |                  ^
[!] Compilation error
```
Thanks for the great work

**Comment by adrien-thierry** [comment]
Hi @InDIOS ,

thank you for this issue, i'm aware of it, NectarJS is already doing this trick at the global scope but not at function scope

I've started to implement it at function scope, but didn't finish yet, 

I will try to fix it soon

**Comment by InDIOS** [comment]
That's a good news. I can help you with some guide lines:
1- You need to have an array to catch identifiers in `VariableDeclaration`, consider all kind of `VariableDeclaration` (ObjectPattern, ArrayPattern, RestParams).
2- You have to replace every `VariableDeclarator` with an `AssigmentExpression`, here is were can be dificult because you need to move the logic of `VariableDeclarator` to `VariableDeclaration`, ObjectPatterns are special because can't use as `AssigmentExpression`. You can sorround it both (id and init properties in `VariableDeclarator`) in an `ArrayPattern` to do it right.
Something like this: `[{ a }] = [{ a: 'a' }]`. Later you have to figure it out how to handle this 😅
3- Consider remove `VariableDeclarator` with null init property except in `for in` and `for of` statements.
4- Put the array of identifiers as a `VariableDeclaration` as firts statement inside function body.

I've made this but using my own AST walker. Maybe giving it to you as and example can help you to get an idea. The example consider all the above options an others like `FunctionExpression` and `ArrowFunctionExpression`

**Comment by adrien-thierry** [comment]
Perfect, thx !

**Comment by InDIOS** [comment]
Here is the example (read comments):
```js
const visitor = require('../visitor'); // My own AST Tree Walker

const parserOptions = { next: true }; // Some parser options (I use Meriyah, it is super fast)
// Testing code
const script = `function some(param) {
  if (param) {
    var a = 1;
  }
  console.log(a);
}
some();`;
// Ids store
const ids = [];
// Let's bigin
const { code } = visitor(script, {
  leave(node) { // this is general, for that reason i check the node type
    // checking if there is ids and if node is correct
    if (ids.length && (node.type.includes('Function') || node.type === 'Program')) {
      const varDec = this.fromCode(`var ${ids.join(', ')}`); // create a VariableDeclaration node
      if (node.type === 'Program') { // global scope
        node.body.unshift(varDec);
      } else if (node.body.type === 'BlockStatement') { // function scope
        node.body.body.unshift(varDec);
      }
      ids.splice(0, ids.length); // reset ids store for reuse
    }
  },
  VariableDeclaration(node) {
    if (node.kind === 'var') { // check only var declarations
      // find the closest parent function or global scope
      const fnParent = this.findParent(p => p && (p.type === 'Program' || p.type.includes('Function')));
      // check if there is a scope and declaration is in that scope
      if (fnParent && fnParent.parent.body !== this.parent) {
        const assign = [];
        for (let i = 0; i < node.declarations.length; i++) {
          const decl = node.declarations[i];
          collectIdentifiers(decl.id, ids); // collect every id node (see below)
          const declId = this.toSource(decl.id); // generate id source
          if (decl.init) { // check for init property
            const declInit = this.toSource(decl.init);
            if (decl.id.type !== 'ObjectPattern') {
              assign.push(`${declId} = ${declInit}`); // make an AssigmenExpression node
            } else {
              assign.push(`[${declId}] = [${declInit}]`);// make an AssigmenExpression node from ObjectPattern
            }
          } else if (/^(For(In|Of))/.test(this.parent.type)) { // check for loops (forIn and forOf)
            assign.push(declId);
          }
        }
        const replace = this.fromCode(assign.join(',')); // make a replacer node
        this.replace(replace?.expression); // replace it
        this.skip(); // skip children (here can goes VariableDelarator logic)
      }
    }
  }
}, { parserOptions });

// helper function that collect id nodes, stolen from acorn-global
function collectIdentifiers(node, ids = []) {
  switch (node.type) {
    case 'Identifier':
      !~ids.indexOf(node.name) && ids.push(node.name);
      break;
    case 'ObjectPattern':
      node.properties.forEach(node => {
        collectIdentifiers(node.value || node.argument, ids);
      });
      break;
    case 'ArrayPattern':
      node.elements.forEach(node => {
        if (node)
          collectIdentifiers(node, ids);
      });
      break;
    case 'RestElement':
      collectIdentifiers(node.argument, ids);
      break;
    case 'AssignmentPattern':
      collectIdentifiers(node.left, ids);
      break;
    default:
      throw new Error(`Unrecognized pattern type: ${node.type}`);
  }
}

console.log(code); // print result
```
I hope it is useful.

**Comment by adrien-thierry** [comment]
Yes, thank you very much, i was looking for this :

`this.findParent(p => p && (p.type === 'Program' || p.type.includes('Function')));` 

Meriyah looks really good, i will check if it could be easy to compile with NectarJS

**Comment by adrien-thierry** [comment]
Thank your very much, corrected in 91a37be
